[bug] WIP fix to deduplicate drilldowns

AnthonyHewins opened the pull request:
This is definitely not done, but if you want to check if this is anywhere in the expected ballpark, please go ahead

jspeis commented:
@hwchen what do you think so far?

The only thing I might suggest adding so far would be a test case for the `Ok` case. Nice work so far @AnthonyHewins !

AnthonyHewins commented on tesseract-core/src/query.rs:
```diff
@@ -490,3 +504,22 @@ impl FromStr for RateQuery {
+    #[test]
+    fn test_validate() {
+        let mut q = Query::new();
+        assert!(q.validate().is_ok());
```
@jspeis is this what you mean? General question too, should I be putting multiple assertions in one function or should I do one assertion per function, which one is better practice in Rust?

jspeis replied:
@AnthonyHewins, something like that -- yes is what I had in mind with two minor notes to the example above exception: to be a valid query a query should have at least one drilldown and at least one measure (cc @hwchen) so technically I would expect the default query which has a empty vec for both drilldowns and measures to be _invalid_. 

In terms of which is a better practice, I think it's more a matter of style and preference. Overall I think it's good to have a test function focus on one task, which tends to make it easier to give a function a descriptive name that others will be able to follow and understand. In this particular case, I could see it going either way because the logic is not particularly complicated.

hwchen replied:
It is a little flexible, but I generally have more than one assertion in a test function. It really depends on the test.

For this series of tests, I might do:
- validate_missing_params (which could include assertions for missing drilldown, missing measure, and both)
- validate_rca
- etc.
Let me know if this clarifies things.

hwchen replied:
And yes, a valid query should have both a drilldown and a measure.

hwchen commented:
Thanks again @AnthonyHewins for your work, really sorry I've been slow to respond, I'm a bit busy at the moment.

AnthonyHewins commented:
Thanks for the comments, I just updated my PR. Added logic to check that measures and drilldowns both exist. Also made the assertions more verbose to make sure it was picking up on the right error.

Apologize if I'm lacking a bit of knowledge in the OLAP domain...I learned about them very briefly about 3-4 years ago and it wasn't in-depth. I will put aside some time to get up to speed, it's a realm I'm going to encounter anyways at some point.

jspeis commented:
@AnthonyHewins this is looking very good! I will take a closer look tomorrow and test it out

jspeis commented:
Looks good to me so far. @hwchen in terms of hooking this up, were you thinking that this would get called in `do_aggregate`?

hwchen commented:
yup @jspeis that sounds good.

jspeis commented:
@AnthonyHewins would you feel comfortable hooking up your `validate` function in the `do_aggregate` functions? (both in the regular and streaming modes?)

We're happy to help if you have any questions!

AnthonyHewins commented:
Absolutely, this weekend is probably when I can get around to doing that. I may have some questions along the way but I'll just post them after I make a commit

jspeis commented:
Great! And feel free to let us know if you have any questions

On Thu, Oct 10, 2019, 12:00 PM adm <notifications@github.com> wrote:

> Absolutely, this weekend is probably when I can get around to doing that.
> I may have some questions along the way but I'll just post them after I
> make a commit
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tesseract-olap/tesseract/pull/195?email_source=notifications&email_token=AAAVMHN7BKKXVUEIM6T7YKTQN5GRRA5CNFSM4IWYYS3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA434NI#issuecomment-540655157>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVMHKR3EFZEVIE4FIFMKTQN5GRRANCNFSM4IWYYS3A>
> .
>

AnthonyHewins commented:
Sorry it's taken some time, things got a little crazy the past two weeks, but I have gotten around to implementing part of it. I want to spend some time with actix so I know how to properly send the right error with `FutureResponse` before I push up though, once I get a chance to reach that part in the tutorial I will push up and hopefully this drawn-out PR will come to a close lol
